packages/swing-store-lmdb/src/lmdbSwingStore.js

```diff
@@ -164,72 +165,212 @@ function makeSwingStore(dirPath, forceReset = false) {
    *
    * @param {string} streamName  The stream to read
+   * @param {Object} startPosition  The position to start reading from
+   * @param {Object} endPosition  The position of the end of the stream
    *
-   * @yields {string} an iterator for the items in the named stream
+   * @returns {Iterable<string>} an iterator for the items in the named stream
    */
-  function* readStream(streamName) {
-    try {
-      const fd = fs.openSync(`${dirPath}/${streamName}`, 'r');
-      if (fd) {
-        const reader = new Readlines(fd);
-        while (true) {
-          const line = /** @type {string|false} */ (reader.next());
-          if (line) {
-            yield line;
-          } else {
-            break;
+  function readStream(streamName, startPosition, endPosition) {
+    insistStreamName(streamName);
+    const status = streamStatus.get(streamName);
+    assert(
+      !status,
+      X`can't read stream ${q(streamName)} because it's already in use`,
+    );
+    insistStreamPosition(startPosition);
+    insistStreamPosition(endPosition);
+    assert(startPosition.itemCount <= endPosition.itemCount);
+
+    let itemCount = endPosition.itemCount;
+    if (endPosition.offset === 0) {
+      assert(itemCount === 0);
+      return [];
+    } else {
+      const filePath = `${dirPath}/${streamName}`;
+      fs.truncateSync(filePath, endPosition.offset);
+      const fd = fs.openSync(filePath, 'r');
+      streamFds.set(streamName, fd);
+      activeStreamFds.add(fd);
+
+      const readStatus = `read-${statusCounter}`;
+      statusCounter += 1;
+      streamStatus.set(streamName, readStatus);
+      // let startOffset = 0;
+      let skipCount = startPosition.itemCount;
+
+      // itemCount -= startPosition.itemCount;
+      // startOffset = startPosition.offset;
+
+      // We would like to be able to seek Readlines to a particular position
+      // in the file before it starts reading.  Unfortunately, it is hardcoded
+      // to reset to 0 at the start and then manually walk itself through the
+      // file, ignoring whatever current position the fd is set to.
+      // Investigation has revealed that giving the Readlines constructor a
+      // 'position' option for where to start reading is a trivial (~4 lines
+      // of code) change, but that would cause us to diverge from the official
+      // npm version.  There are even a couple of forks on NPM that do this,
+      // but they have like 2 downloads per week so I don't trust them.  Until
+      // this is resolved, the only way to realize a different starting point
+      // than 0 is to simply ignore records that are read until we catch up to
+      // where we really want to start, which the following code does.  It's
+      // not ideal, but it works.
+      //
+      // const innerReader = new Readlines(fd, { position: startOffset });
+      const innerReader = new Readlines(fd);
+      function* reader() {
+        try {
+          while (true) {
```

Comment:
nit: collapse this into the assert, `prettier` means we've already paid the cost of smearing this out across four lines, might as well make one of them wider and remove the spurious `statusInner` name:
```js
while (true) {
  assert(
    streamStatus.get(streamName) === readStatus,
    ...
  );
```

Reply:
I think this was an artifact of debugging, where I was watching what it was doing, but not needed now.